Add semantic markup to ASCII Doc source

PaulMartinsen opened the pull request:
> [!IMPORTANT]
>
> [2025-09-12-sdpi-supplement.zip](https://github.com/user-attachments/files/22293710/2025-09-12-sdpi-supplement.zip) containing the processed document is attached to this PR because the github action, which would normally generate this content, uses the processor from the master branch missing out on the tweaks. 


## 📑 Description &mdash; first pass

This is ready for comments/feedback on adding semantic markup to ASCIIDoc source for exporting requirements, profiles etc. It includes:

* profiles and profile options including:
  * transaction references
  * content module references
  * use-case references
  * actor markup
* requirements
  * markup normative, note, examples, etc content
  * proposed requirement types (tech feature, risk-mitigation, etc)
* gateways (for requirement 'ownership')
* use cases
  * section
  * background
  * scenarios
  * steps
 * oids:
   * actors
   * transactions (defaults to appending `transaction-number` and `transaction-number.1` to SDPi transaction root arc `1.3.6.1.4.1.19376.1.6.4`)
   * profiles
   * use-cases
   * content modules
   * requirements

The new markup enables automatic generation of summary tables, which are included for consideration:
* profile actors and transactions
* profile actors and content modules
* implementation conformity tables for each actor (in volume 4 until we decide if these are useful and where they should go)
* oid summary for actors and transactions (in volume 4 until we decide if they are useful and where they should go)


I noticed and fixed some problems in the text:
* some bibliography entries weren't formatted correctly, with an extra comma following the bibliography anchor
* the same anchor id, `vol2_clause_dev_23_message_announce_network_presence_ses`, was used in `tf2-dev-25.adoc`, `tf2-dev-23.adoc`. 

The document processor has a few tweaks as well, including:
* reverted [PF443](https://github.com/IHE/DEV.SDPi/pull/443) change so appendicies link text is correct ([issue #453](https://github.com/IHE/DEV.SDPi/issues/453)), though this may also restore  the problem #443 wanted to resolve,
* changed reference tuple separator from a random GUID to a '\*' so that running the document processor twice produces identical output each time,
* added support to use an actor as a filter when generating ICS tables,
* added support for actor options, needed for some profiles,
* added CSS class to selected links so tables use transaction names instead of section numbers,
* added support to "include" deferred transactions and content modules in profiles,
* inherit table titles and ids when inserting placeholders for ICS, transaction and content module tables,
* improved detection of bibliography entries,
* more robust parsing of actors from requirement normative statements,
* more robust parsing of section titles and multi-line actor definitions.

## ☑ Mandatory Tasks

The following aspects have been respected by the pull request assignee and at least one reviewer:

- Changelog update (necessity checked and entry added or not added respectively)
  * [ ] Pull Request Assignee
  * [ ] Reviewer

## For reviewers

 [SDPi requirements.xlsx](https://blueleafsoftware.sharepoint.com/:x:/s/2023-FPH-SDC/ESbtkdZL8K9CvpgpX0gKM4YBw0IZr_ExFdIg0BRxVtelgA?e=8f8IQ8) has a list of all requirements, profiles and use case, which may be useful for checking. Updating the spreadsheet won't update the ASCIIDoc source, unfortunately; please update the source or post a comment below with the changes needed.

* [ ] Check requirements have been typed correctly.
  * Types were assigned as a starting point; one or all may be wrong,
  * The "Requirements" sheet lists all requirements,
  * The "Risk-Requirements" sheet lists all risk-mitigation requirements, along with a guess at [ses type and testability](https://profiles.ihe.net/DEV/SDPi/index.html#vol1_clause_sdpi_requirement_type_ses),
  * (currently) all requirements seem to be either *tech features* or `risk mitigation`, 
  * 
* [ ] Check profile transactions
  * The "Profile-all transactions" sheet lists the transactions associated with all profile, profile and actor options 
  * The "Profile transactions" sheet list transactions with their actors, contribution and obligation
  * The "Transactions-ActorOptions" sheet lists transaction contribution and obligations for actor options.
  * The "Transactions-ProfileOptions" sheet lists obligations for transaction contributions in profile options. 
* [ ] Check use cases
  * The "Profile use cases" sheet lists all the profiles and their actor's obligations to use cases. 
* [ ] Check profile actors and transactions tables (e.g., *Table 1:10.1-1. SDPi-P Profile - Actors and Transactions*) and profile actors and content modules tables (e.g., *Table 1:10.1-3. SDPi-P — Actors and Content Modules*)
  * The original tables are included, unchanged, for comparison
  * Maching generated formatting is slightly different but the same information is (should be!) presented

## Styling

There are some minor changes to document styling arising from the markup. Currently I've left everything with the default ASCIIDoc styling. It looks okay to me but this can be changed if it isn't suitable. See examples below.

* Transactions
  * the normative statement ends up in a yellow box
  * notes are not collapsed (this was really annoying as the notes usually contain really helpful information)
  * examples and related content are collapsed
* The structured gherkin statements for use cases ended up in a yellow box
* The actor roles table in transactions got an extra column for the actor's contribution since they are defined here anyway.
* The profile actors and transactions, and content modules tables are a bit different, as mentioned above. 

<img width="1233" height="363" alt="image" src="https://github.com/user-attachments/assets/fc222cd7-1217-44a9-85b3-660c013033ca" />

<img width="1246" height="441" alt="image" src="https://github.com/user-attachments/assets/f222ed39-c0f3-40f3-968c-570ae537fa55" />

<img width="1245" height="231" alt="image" src="https://github.com/user-attachments/assets/4464a04a-cece-44e4-85ca-1a0d653bb24e" />

d-gregorczyk commented on asciidoc/volume2/gateways/tf2-ch-b-gateway-obx4-mapping.adoc:
```diff
@@ -2,34 +2,36 @@
 
 .R8014
-[sdpi_requirement#r8014,sdpi_req_level=shall,sdpi_max_occurrence=1]
+[sdpi_requirement,sdpi_req_level=shall,sdpi_req_type=tech_feature]
```
This needs to be reverted. The Asciidoc id is required in order to be able to increase compatibility with unprocessed Asciidoc (e.g. when using preview). Removing the id from the source kills every cross-reference from the document. I did this redundant modelling for a reason ;-)

PaulMartinsen replied:
That's easy enough.

Just the ids or the `sdpi_max_occurance` as well? I couldn't figure out what `sdpi_max_occurance` was for and it was never anything except 1. If `sdpi_max_occurance` is necessary, perhaps it could default to reduce clutter in the source?

My question is where do OIDs fit in: would we want links, that folk potentially copy, to be to OIDs or just the local ids? Using inline macros (e.g., `RefRequirement:R1021[]`) lets us decide later and might be cleaner for the json export but also may not be preview friendly enough, amongst other things. Maybe making things easier for editors is more important.

PaulMartinsen replied:
I sorted out the IDs, so every requirement has both the title and anchor as before. 

@PeterKranich , can you shed any light on whether `sdpi_max_occurance` is still needed and what its for? 

All instances had the value one. If it is still needed, perhaps it could default to 1 if not present in the source?

PaulMartinsen commented:
> **2025.08.15 SDPi Friday** - @PaulMartinsen ... can you give us a status update on this PR?

I've completed most of the steps from the roadmap in #463 for this iteration. Things left to resolve are:
* are the oids handled correctly, pending clarification in #449 ,
* clarification of `sdpi_max_occurance` item above,
* addressing things arising from reviewer notes above,
* resolving any other comments,
* collectively deciding what to do with the content volume 4 placeholder & automatically generated tables.

JavierEspina commented:
Looking at the ZIP pasted in Paul´s first comment I would say that:

- the styling differences are acceptable
- There is an error in the type column in the table in 4:1.1 Device integration profiles. It should say Profile not Transaction.
-

PaulMartinsen commented:
> Looking at the ZIP pasted in Paul´s first comment I would say that:
> 
> * the styling differences are acceptable
> * There is an error in the type column in the table in 4:1.1 Device integration profiles. It should say Profile not Transaction.

Oops. Good catch; thanks @JavierEspina . Pushed fix and uploaded zip .

JavierEspina commented:
> * collectively deciding what to do with the content volume 4 placeholder & automatically generated tables.

**SDPi Call Aug 22:**
Volume 4 is reserved for somethign else. So, the content needs to go elsewhere. Specifically:

1. "4:1 OID identifiers" should become Appendix 3:C (i.e. in Volume 3)
2. "4:2 Implementation conformity statements" should move to another SDPi document. Yet TBD what kind of document, but probably something like the generated requirements JSON file. Todd will discuss this directly with Paul.

ToddCooper commented:
@PaulMartinsen  ...

> I moved both to appendix 3, C for the oids and D for the ICS for this PR. Shifting the ICS to a separate document may be better handled by a separate PR.

Thanks, Paul.  I was thinking that ultimately, we may just want to push this to TF-3Z since my assumption is that we will always want it to be the last in the list.  Yes?  

For now, we can keep the SDPi ICS as TF-3D, especially as we evaluate what is most helpful in the long run.  And we may want both!  Many standards (IEEE 11073 especially) have ICS tables in the back to aggregate requirements under specific "topics" or functional aspects and to state how conformance can be claimed, esp. stating what was NOT implemented, extensions, etc.  

FWIW - IHE Profile ICS tables is a totally new thing.  It will be huge and I believe well received, but also new ground.  This is totally different from the style guides for TF-1 to -3, where there is established and strongly held precedent.  So tweaking those Actor/Transaction table formats, for example, is NOT for the faint of heart!  And not to be done without serious consideration.

JavierEspina commented:
> I think all the issues are resolved in this one now and it is good to go/ another review.
> 
> Note
> 
> I have included oids on requirements, content modules and use-cases using the [new root oids](https://github.com/IHE/DEV.SDPi/issues/449#issuecomment-3237155837). Do we need to wait for domain approval before merging this?

Thanks Paul. We need to review this one now that it is ready for prime time. Note that there is a dependency with new issue #474. But back to your question: earlier this week, in Silver Spring, MD, the IHE DEV agreed to go ahead with those .8, .9 and .10 arcs for the Domain. So we are fine with that.

PaulMartinsen commented:
Thanks for the info on oids @JavierEspina . I've ticked off the OID item  on my to do list :)

It will be good to get this one merged so new requirements &c can use the new markup as well. And once its merged I can make a start on incorporating the 3rd party standards references as well. Using 10700 as a test case. 

Perhaps this PR may not need to wait on #474? #474 might depend on #449 though since we still need to figure out a plan for actor and profile option oids; I guess this is part of the #449 goals. In any case, I updated the requirement OID text to match where we landed on oids for those (see my last comment in #449) 

BTW, github shows me a "Changes requested" note, but I believe I've addressed all the requests; at least I can't see any unresolved ones. Do you see any I missed?

PaulMartinsen commented:
@ToddCooper , @JavierEspina , what do you think about splitting the gateway direction option? That is instead of having one option with an enumeration, there would be two options: `gateway-direction-import` and `gateway-direction-export`. 

Two reasons:
* If we want to indicate the presence of options with oids in X509 certificates, could that be through the extended key usage? x.509 certificates don't allow the oid in an EKU  I believe, so it will be easier to map options => oids if they are separate. Assuming this is something we want to do without defining a namespace extension for certificates? 
* Current markup supports profile options but not conditional values for them. Will this be the only example of an option with, effectively, sub-options? What others can we expect to encounter? 

It also seems, to me at least, that a gateway supporting importing and exporting are very different. For example, an importing gateway is as different to an exporting gateway as it is to the manged proxy option. Is there a danger we're confusing actor and option nomenclature here? That is, the actor is "gateway" but the option is "import" or "export". Though I realize these are profile options, not actor options.

PaulMartinsen commented:
I'm a bit confused by `SOMDS v2 gateway` and `SOMDS FHIR gateway` and `SOMDS consumer` too. Particularly that the `SOMDS consumer` can be a `SOMDS consumer` or a `SOMDS provider`, but these are quite different things.

It seems like there are `SOMDS gateway consumer` and a `SOMDS gateway provider` actors, which would connect SDC data to an external network or provide data from an external network respectively. And that the `SOMDS v2 gateway` and `SOMDS FHIR gateway` aren't separate actors but implementations of gateways (consumer or provider or both) because their entries in the "Table 1:10.1-2. SDPi-P Profile - Actors and Transactions" table are just the same. More like use cases even? 

Are `SOMDS v2 gateway` and `SOMDS FHIR gateway` could be actors with other transactions outside SDPi? So not use cases? 

Anyway, I wonder if we need to address the question "what transactions do I need to support if I want to implement an exporting gateway?" more clearly? 

Am I missing something here?

PaulMartinsen commented:
Completed next steps:
* switched back to manual actor and transactions tables (with document attribute `markup-actor-transactions-table` set in `sdpi-supplement.adoc`),
* merged profile options and actor options because they are not different things,
* proposed alternative rendition for gateway direction, splitting import and export into separate (boolean) options. 

Discussed the last, in particular, with @ToddCooper  last week, but he'll need to check it panned out okay. If too horrific, I can incorporate changes or go back to option with metadata (that is, one gateway direction option + import/export/bidirectional as option attribute) and investigate some representation for this in the semantic markup. 

> [!IMPORTANT]
> [2025-10-06-sdpi-supplement.zip](https://github.com/user-attachments/files/22718070/2025-10-06-sdpi-supplement.zip)  containing the processed document is attached because the github action, which would normally generate this content, uses the processor from the master branch missing out on the tweaks in this PR.

Note: this PR is still showing the following message, but there don't seem to be any actual requested changes that I can find . 

<img width="404" height="67" alt="image" src="https://github.com/user-attachments/assets/38eb880d-834f-4ab5-82e0-cce47b8a98ff" />

d-gregorczyk commented on .ci/asciidoc-converter/src/test/resources/req-ownership.adoc:
```diff
@@ -4,13 +4,13 @@
 Pellentesque pellentesque ligula vitae neque porttitor sollicitudin. Nullam at nisi nunc. 
 
-[role="profile",profile-id="Profile-A"]
+[role="profile",profile-id="Profile-A",oid-arcs=.99]
```
Is this meant to be a CSS class name? `role` is a reserved attribute in Asciidoc. We need to prefix every attribute specific to SDPi with `sdpi_` because of that - or make sure it is not reserved by Asciidoc.

d-gregorczyk replied:
I would also tend to stick to the underscore notation used for other attributes, i.e., `sdpi_oid_arcs` or `oid_arcs`.

PaulMartinsen replied:
Role here is the asciidoc attribute. We leverage their system for simplicity. I've avoided all asciidoc reserved words in attribute names. I stopped using the sdpi_ prefix because it gets very distracting in the source for editors. All attributes are scoped to the block so there's no real need for scoping by convention.

PaulMartinsen replied:
I think we cleared this up in email exchange. Rationalization of naming conventions will happen in part 4 of #463

JavierEspina commented:
Just did another review (superficial by necessity!) of the generated artifact. It looks good to me as well. Only one question: I saw that R0800, R0801and R0802 are shall level requirements that, accordingly, show up in the ICS table for a SOMDS provider. Should that be conditional to the Provider/Consumer using the sdpi:ReportSequence extension (which is "should" in all 2:A.5.2 requirements)?

JavierEspina reviewed (changes requested):
See my two 16-October comments in the conversation

PaulMartinsen commented on CHANGELOG.md:
Good catch. I updated the change log.

JavierEspina commented:
Just approved, but it is still blocked by a David Change Request - probably that one we could not locate any more... Going through the flies I found:
<img width="1538" height="694" alt="image" src="https://github.com/user-attachments/assets/4008ddd7-79e7-4ab1-97d5-bd1bc0b120ea" />
@d-gregorczyk , can it be marked as resolved?
